### DISCUSSION TITLE
Better block parameters?

### DISCUSSION THREAD

**Discussion by maxsupermanhd** [Tnze/go-mc#157] [Ideas]
I was thinking about better way of storing block state information and I could not understand why can not it be a regular `map[string]string`. Also speaking of current implementation, it lacks method to get all the possible values of a field.

PS I am trying to reconstruct 3d model of things so I am digging into jar, blockstates, models and textures pretty hard here

**Comment by Tnze** [comment]
The current design is optimized for performance that block state serialization. The map is a horrible choice in this case. It takes a lot more memory but offers a very low operating speed. Regardless of heap memory allocation overhead, it cannot even be easily compared! 

For example, how do you convert a `map[string]interface{}` as block state into state id? Take all properties, sort them by keys, calculate hash, then map them to state id? There is an implementation in here https://github.com/df-mc/dragonfly and you can take a look if you want.

**Reply by maxsupermanhd** [reply]
Well, actually, I only want to have an easy access to list of all possible parameter values (with getter, just like String method). I am struggling to build blockstate description converter to actual block structure because I don't know all possible values that parameter can take, some descriptions only define some fields, others must be considered in all variations.
Here is what I have now:
```

func getBlockFromDescription(name, description string) block.Block {
	if block.FromID[name] == nil {
		return nil
	}
	if description == "" {
		return block.FromID[name]
	}
	params := map[string]string{}
	for _, d := range strings.Split(description, ",") {
		kv := strings.Split(d, "=")
		if len(kv) != 2 {
			panic("kv pair is not a pair")
		}
		params[kv[0]] = kv[1]
	}
	bp := reflect.New(reflect.ValueOf(block.FromID[name]).Type())
	b := bp.Elem()
	for m := 0; m < b.NumField(); m++ {
		fieldName := b.Type().Field(m).Name
		methodUnmarshal := -1
		for methodI := 0; methodI < b.Field(m).Addr().NumMethod(); methodI++ {
			if b.Field(m).Addr().Type().Method(methodI).Name == "UnmarshalText" {
				methodUnmarshal = methodI
				break
			}
		}
		if methodUnmarshal == -1 {
			log.Printf("Failed to find method UnmarshalText in type [%s]", b.Field(m).Type().Name())
			continue
		}
		for k, v := range params {
			if k == strings.ToLower(fieldName) {
				b.Field(m).Addr().Method(methodUnmarshal).Call([]reflect.Value{reflect.ValueOf([]byte(v))})
				break
			}
		}
	}
	return b.Interface().(block.Block)
}
```
Code above takes block id and state description from jar assets and sets them accordingly however such function should return array of all possible blockstates that match description, currently impossible without looping over and checking states id array. I would really want to avoid performing array lookup.

**Reply by Tnze** [reply]
Maybe I don't fully understand your purpose. But why don't you traversal the `StateList`? And the same block 's state id will be adjacent.

**Reply by maxsupermanhd** [reply]
Because traversing StateList is less efficient than just making a block generator that directly creates block array with defined parameters

**Reply by Tnze** [reply]
Ok , so what is the 'description'? any example?

**Reply by maxsupermanhd** [reply]
Block state description is this variant string in `/assets/minecraft/blockstates/` jsons, described as following: `facing=east,half=lower,hinge=left,open=false` (acacia\_door.json)

**Reply by maxsupermanhd** [reply]
Those variants dictate what model is being rendered with what texture and with what geometry

**Reply by Tnze** [reply]
Can we do that by improve the block data generator to add additional `Model()` method?

**Reply by maxsupermanhd** [reply]
Yes, that would be amazing

**Reply by Tnze** [reply]
Then before I add them, you can generate some []Model, mapping StateID to Model, can't you?

**Reply by maxsupermanhd** [reply]
I can wait for you to make them, not urgent. I am still trying to migrate WebChunk to new blockstates and it is complete pain

**Reply by Tnze** [reply]
I'm confused about what you actually want to do. 

Why you must access to list of all possible parameter values?

You said you want convert description to actual block structure ('s array?). But block.Block and StateID are one-to-one. Do you need a `map[description][]StateID`?

**Reply by Tnze** [reply]
Or do you want a reverse mapping?

**Reply by maxsupermanhd** [reply]
I need to map every state id to a color from model/texture to have a color palette to render a map.
I also need a page where I can edit every color by hand like I want for example to have powered lamp to be slightly brighter.
Have ability to encode it in gob too but I guess it works

**Reply by maxsupermanhd** [reply]
Also have a model definition loaded for every state id to have ability to potentially render everything in 3D

**Reply by Tnze** [reply]
> some descriptions only define some fields, others must be considered in all variations.

This can't be high performance, right? Can we avoid doing this at runtime?

**Reply by maxsupermanhd** [reply]
Yes, it will be done on init at worst, but it have to be prepared for every block state.
Rules are defining *some* blockstates, for example it can define facing but not power state that means you have to generate all blockstates with power state to be both on and off

**Reply by Tnze** [reply]
Can you traversal StateList and generate this mapping? Sounds like it's unnecessary to add them as Block's method now:)

**Reply by maxsupermanhd** [reply]
At this point I am doubting I will continue WebChunk development on the go-mc base because you have stuff changing all the time, I have finite resources afterall. Generating color palette is *hard* with block states being considered, I might ditch whole project because of that change. Today you have one thing tomorrow you will have another, I will better stick to old version of go-mc or move out completely to my own model just to simplify everything

**Reply by Tnze** [reply]
Yes, you have to make choices in development. Use go-mc or not. It depends on your needs.

I make choices too.
